core/include/traccc/finding/details/debug_output.hpp

```diff
@@ -0,0 +1,28 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2023-2025 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+#include "traccc/utils/logging.hpp"
+
+namespace traccc {
+enum class propagate_to_next_surface_exit_mode {
+    HOLE_LIMIT_REACHED,
+    BRANCH_LIMIT_REACHED,
+    MAX
+};
+
+struct propagate_to_next_surface_debug {
+    void reset();
+
+    unsigned int m_failure_count[static_cast<std::size_t>(
+        propagate_to_next_surface_exit_mode::MAX)];
+};
+
+void log_propagate_to_next_surface_debug(
+    const propagate_to_next_surface_debug& obj, const Logger& logger);
```

Comment:
I wonder what our convention should be for printing custom types. 🤔

I believe if you define

```c++
std::ostream& operator<<(std::ostream&, const propagate_to_next_surface_debug&);
```

, that could be used to print such objects to a `traccc::Logger` (`Acts::Logger`) object. Like we do here for instance:
  - https://github.com/acts-project/traccc/blob/main/performance/include/traccc/performance/timing_info.hpp#L39
  - https://github.com/acts-project/traccc/blob/main/performance/include/traccc/performance/throughput.hpp#L37
  - https://github.com/acts-project/traccc/blob/main/examples/run/common/throughput_st.ipp#L214-L225

Reply:
In general, defining `operator<<` for a type is an antipattern unless there is a single, ubiquitous way to turn a value of the type into a string. In this case, there are many different wordings and ways to print this data, so `operator<<` is a no-go. Of course, we can debate that there might be better names for this function.

Reply:
I think it's pretty safe to define an operator for the "standard way" of printing the object. Which will also allow clients to use the code more flexibly. Printing not just to a logger, but to any kind of stream. And also not just at a `DEBUG` level, but also any other that they want. (Very realistically we may end up using such a printout in a warning or error message as well. In case the debug information points at some issue.)

So this I feel fairly strongly about.

Reply:
Consider what happens if we want to print some exit modes at a different severity level than others; then your approach breaks down. I think it's much better to have a function which calls the logger itself, rather than producing a string for the whole object.